Alpine Linux based images

jkawamoto opened the pull request:
It'd be nice if there are images based on Alpine Linux.

jkawamoto committed:
Adding Dockerfile to build images based on Alpine Linux

ro-tex commented on alpine/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+ARG SIA_DIR="/sia"
+ARG SIA_DATA_DIR="/sia-data"
+
+RUN mkdir /lib64 && ln -s /lib/libc.musl-x86_64.so.1 /lib64/ld-linux-x86-64.so.2
```
REQ: Can you combine this `RUN` with the one on line 25? And maybe split the commands on separate lines similar to the `RUN` on line 11?

jkawamoto replied:
Fixed in 6a9185c.

ro-tex commented on alpine/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+  ./siad \
+    --modules "$SIA_MODULES" \
+    --sia-directory "$SIA_DATA_DIR" \
+    --api-addr "localhost:8000"
```
```suggestion
COPY run.sh ./
ENTRYPOINT ["./run.sh"]
```

REQ: We just switched away from using the shell version of `ENTRYPOINT` because it prevented `siad` from performing a graceful shutdown. Can you please use the `run.sh` script as entrypoint? You'll also need to copy it here from the project's root or we can figure out a less WET approach.

jkawamoto replied:
Fixed in 6a9185c.

ro-tex reviewed (commented):
Looks good! We only need to bring it up to speed with the other Dockerfiles and we're good to go.

jkawamoto committed:
Improve Dockerfile

as requested in
* https://github.com/NebulousLabs/docker-sia/pull/4/files#r409685997
* https://github.com/NebulousLabs/docker-sia/pull/4/files#r409687457

jkawamoto committed:
Add a HEALTHCHECK

jkawamoto committed:
Use /bin/sh in shebang

jkawamoto commented:
Thank you for your review. I modified Dockerfile and add the health check introduced in #8 as well. 

One problem is that Alpine Linux doesn't have bash by default. I, hence, modified shebang to use OS's default shell. Is it an acceptable change? If not, I'll copy shell scripts in alpine directory so that the main image can use `/usr/bin/env bash` and the alpine one can use `/bin/sh`.

ro-tex commented on alpine/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+      unzip -j "$SIA_ZIP" "${SIA_PACKAGE}/siad" -d /sia
+
+FROM alpine:3
+LABEL maintainer="NebulousLabs <developers@nebulous.tech>"
```
```suggestion
LABEL maintainer="NebulousLabs <devs@nebulous.tech>"
LABEL autoheal=true
```
If you're adding the HEALTHCHECK you might as well label the container for autohealing. We also changed our email (there is an open PR).

jkawamoto replied:
Fixed in e209584.

ro-tex commented on alpine/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+FROM debian:stretch-slim AS zip_downloader
+LABEL maintainer="NebulousLabs <developers@nebulous.tech>"
```
```suggestion
LABEL maintainer="NebulousLabs <devs@nebulous.tech>"
```

jkawamoto replied:
Fixed in e209584.

ro-tex commented on alpine/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+ENV SIA_MODULES gctwhr
+
+COPY healthcheck.sh .
+COPY run.sh ./
```
```suggestion
COPY run.sh .
```
Let's keep it consistent. :)

jkawamoto replied:
Fixed in e209584.

ro-tex reviewed (commented):
Switching from `/bin/bash` to `/bin/sh` is perfectly fine but that won't help you avoid the copying of the shell scripts. Docker won't allow you to build from `/alpine/Dockerfile` and include files from `/`. That's why the `/dev` image also has its own copy of the `run.sh`.

Did you manage to build your image? For me it fails with `COPY failed: [...]/healthcheck.sh: no such file or directory` which is not surprising, having in mind that we don't have a copy in the working directory. Maybe you happened to have a copy there and didn't commit it?

Anyway, we'll need copies of those scripts in the `/alpine` directory. I've added a couple more comments to the PR, so we can bring this image in line with the other ones.

jkawamoto committed:
Update email address.

in accordance with #10.

jkawamoto commented:
Email addresses and the COPY step are fixed.

As for the shell scripts, if you build with this command at the root of this repository, it should work:
```sh
$ docker build -t sia:1.4.7-alpine -f alpine/Dockerfile .
```
(and `docker build -t sia:1.4.7-dev -f dev/Dockerfile .`, too)

ro-tex reviewed (approved):
Looks good and works on my machine. Thanks!
